File: 5week/근우/2.find-leaves-of-binary-tree.py
## https://leetcode.com/problems/find-leaves-of-binar# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import defaultdict, deque


class Solution:
    def findLeaves(self, root: Optional[TreeNode]) -> List[List[int]]:
        ref = defaultdict(list)
        def dfs(node):
            if not node:
                return 0
            if node:
                cur = 1+max(dfs(node.left),dfs(node.right))
            ref[cur].append(node.val)
            return cur
        dfs(root)
        return [v for v in ref.values()]

File: 5week/남규/2. Find Leaves of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool removeLeafs(TreeNode* curr, vector<int>& leafs){
        if(!curr->left && !curr->right){
            leafs.push_back(curr->val);
            return true;
        }
        
        if(curr->left){
            bool isLeaf = removeLeafs(curr->left, leafs);
            if(isLeaf){
                curr->left = NULL;
            }
        }
        
        if(curr->right){
            bool isLeaf = removeLeafs(curr->right, leafs);
            if(isLeaf){
                curr->right = NULL;
            }
        }
        return false;
    }
    
    vector<vector<int>> findLeaves(TreeNode* root) {
        vector<vector<int>> ans;
        if(!root) return ans;
        
        while(root){
            if(!root->left && !root->right){
                ans.push_back({root->val});
                break;
            }
            
            vector<int> leafs;
            removeLeafs(root, leafs);
            ans.push_back(leafs);
        }
        return ans;
    }
};
